src/data/fintype/basic.lean

```diff
@@ -371,12 +385,54 @@ multiset.card_map _ _
 theorem card_congr {α β} [fintype α] [fintype β] (f : α ≃ β) : card α = card β :=
 by rw ← of_equiv_card f; congr
 
+section
+
+variables [fintype α] [fintype β]
+
+/-- If the cardinality of `α` is `n`, there is computably a bijection between `α` and `fin n`.
+
+See `fintype.equiv_fin_of_card_eq` for the noncomputable definition,
+and `fintype.trunc_equiv_fin` and `fintype.equiv_fin` for the bijection `α ≃ fin (card α)`.
+-/
+def trunc_equiv_fin_of_card_eq [decidable_eq α] {n : ℕ} (h : fintype.card α = n) :
+  trunc (α ≃ fin n) :=
+(trunc_equiv_fin α).map (λ e, e.trans (fin.cast h).to_equiv)
+
+
+/-- If the cardinality of `α` is `n`, there is noncomputably a bijection between `α` and `fin n`.
+
+See `fintype.trunc_equiv_fin_of_card_eq` for the computable definition,
+and `fintype.trunc_equiv_fin` and `fintype.equiv_fin` for the bijection `α ≃ fin (card α)`.
+-/
+noncomputable def equiv_fin_of_card_eq {n : ℕ} (h : fintype.card α = n) :
+  α ≃ fin n :=
+by { letI := classical.dec_eq α, exact (trunc_equiv_fin_of_card_eq h).out }
+
+/-- Two `fintype`s with the same cardinality are (computably) in bijection.
+
+See `fintype.equiv_of_card_eq` for the noncomputable version,
+and `fintype.trunc_equiv_fin_of_card_eq` and `fintype.equiv_fin_of_card_eq` for
+the specialization to `fin`.
+-/
+def trunc_equiv_of_card_eq [decidable_eq α] [decidable_eq β] (h : card α = card β) :
+  trunc (α ≃ β) :=
+(trunc_equiv_fin_of_card_eq h).bind (λ e, (trunc_equiv_fin β).map (λ e', e.trans e'.symm))
+
+/-- Two `fintype`s with the same cardinality are (noncomputably) in bijection.
+
+See `fintype.trunc_equiv_of_card_eq` for the computable version,
+and `fintype.trunc_equiv_fin_of_card_eq` and `fintype.equiv_fin_of_card_eq` for
+the specialization to `fin`.
+-/
+noncomputable def equiv_of_card_eq (h : card α = card β) : α ≃ β :=
+by { letI := classical.dec_eq α, letI := classical.dec_eq β,
+     exact (trunc_equiv_of_card_eq h).out }
+
+end
+
 theorem card_eq {α β} [F : fintype α] [G : fintype β] : card α = card β ↔ nonempty (α ≃ β) :=
-⟨λ h, ⟨by classical;
-  calc α ≃ fin (card α) : trunc.out (equiv_fin α)
-     ... ≃ fin (card β) : by rw h
-     ... ≃ β : (trunc.out (equiv_fin β)).symm⟩,
-λ ⟨f⟩, card_congr f⟩
+⟨λ h, ⟨equiv_of_card_eq h⟩,
```

Comment:
```suggestion
⟨λ h, ⟨equiv_of_card_eq h⟩,
```

Reply:
There are no changes in your suggestion. Did you mean:
```suggestion
⟨λ h, by { haveI := classical.prop_decidable, exact (trunc_equiv_of_card_eq h).nonempty },
```

Reply:
I think I did not mean to submit that comment! Your version is probably what I meant to say.